xdsl/backend/riscv/prologue_epilogue_insertion.py

```diff
@@ -0,0 +1,92 @@
+from dataclasses import dataclass, field
+
+from xdsl.builder import Builder
+from xdsl.dialects import builtin, riscv, riscv_func
+from xdsl.dialects.riscv import (
+    IntRegisterType,
+    Registers,
+    RISCVRegisterType,
+)
+from xdsl.ir import MLContext
+from xdsl.passes import ModulePass
+
+
+@dataclass(frozen=True)
+class PrologueEpilogueInsertion(ModulePass):
+    """
+    Pass inserting a prologue and epilogue according to the RISC-V ABI.
+    The prologues and epilogues are responsible for saving any callee-preserved
+    registers.
+    In RISC-V these are 's0' to 's11' and 'fs0' to `fs11'.
+    The stack pointer 'sp' must also be restored to its original value.
+
+    This pass should be run late in the pipeline after register allocation.
+    It does not itself require register allocation nor invalidate the result of the
+    register allocator.
+    """
+
+    name = "riscv-prologue-epilogue-insertion"
+    xlen: int = field(default=4)
+    flen: int = field(default=8)
+
+    def _process_function(self, func: riscv_func.FuncOp) -> None:
+        # Find all callee-preserved registers that are clobbered. We define clobbered
+        # as it being the result of some operation and therefore written to.
+        used_callee_preserved_registers = {
+            res.type: None
+            for op in func.walk()
+            if not isinstance(op, riscv.GetRegisterOp | riscv.GetFloatRegisterOp)
+            for res in op.results
+            if res.type in Registers.S or res.type in Registers.FS
+        }
```

Comment:
Hopefully this will give the same behaviour:
```suggestion
        used_callee_preserved_registers = orderedset(
            res.type: None
            for op in func.walk()
            if not isinstance(op, riscv.GetRegisterOp | riscv.GetFloatRegisterOp)
            for res in op.results
            if res.type in Registers.S or res.type in Registers.FS
        )
```